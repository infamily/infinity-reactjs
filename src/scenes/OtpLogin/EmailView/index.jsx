import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import { Link } from 'react-router-dom';
import CheckBox from 'components/CheckBox';
import ifIcon from 'images/if.png';
import otpService from 'scenes/OtpLogin/services';
import { getErrorMessage } from '../helpers';
import messages from '../messages';

const initialState = {
  captcha_1: ''
};

export default class EmailView extends Component {
  constructor() {
    super();
    this.state = {
      email: '',
      captcha_0: '',
      captcha_1: '',
      captcha: {
        key: '',
        image_url: ''
      },
      tosAgreement: false,
      tosError: false
    };
  }

  static propTypes = {
    view: PropTypes.string.isRequired,
    setPopUp: PropTypes.func.isRequired,
    changeView: PropTypes.func.isRequired,
    changeEmail: PropTypes.func.isRequired
  };

  componentWillMount() {
    if (!this.state.captcha.key) this.refresh();
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.ignoreCaptcha) {
      this.setState({ captcha_1: '123' });
    }
  }

  refresh = async e => {
    if (e) e.preventDefault();
    try {
      const data = await otpService.getCaptcha();
      this.updateCaptcha(data);
    } catch (err) {
      this.setPopUp({
        title: <FormattedMessage {...messages.serverErrTitle} />,
        text: err
      });
    }
  };

  updateCaptcha(data) {
    if (data.key) {
      const image_url = otpService.getImage(data);

      this.setState({
        captcha_0: data.key,
        captcha_1: '',
        captcha: {
          key: data.key,
          image_url
        }
      });
      this.props.changeView('email');
    }
  }

  setPopUp = params => {
    this.props.setPopUp(params);
  };

  onEmailSubmit = async e => {
    e.preventDefault();

    try {
      const { email, captcha_0, captcha_1, tosAgreement } = this.state;
      if (!email || !captcha_1) {
        this.setPopUp({
          title: <FormattedMessage {...messages.signInErrorTitle} />,
          text: <FormattedMessage {...messages.allFilledOut} />
        });
        return;
      }

      if (!tosAgreement) {
        this.handleToSError();
        return;
      }

      const captcha = {
        hashkey: captcha_0,
        response: captcha_1
      };

      this.props.changeEmail(email);
      const params = { email, captcha };
      await otpService.signUp(params);

      this.setState({ ...initialState });
      this.props.changeView('login');
    } catch (error) {
      if (error.response !== undefined) {
        const formattedText = getErrorMessage(error.response.data);
        this.setPopUp({
          title: <FormattedMessage {...messages.signInErrorTitle} />,
          text: <FormattedMessage {...formattedText} />
        });
      }
    }
  };

  handleChange = e => {
    this.setState({
      [e.target.name]: e.target.value
    });
  };

  handleToS = () => {
    this.setState(prevState => ({
      tosAgreement: !prevState.tosAgreement,
      tosError: false
    }));
  };

  handleToSError = () => {
    this.setState({
      tosError: true
    });
  };

  render() {
    const {
      email,
      captcha_0,
      captcha_1,
      captcha,
      tosAgreement,
      tosError
    } = this.state;

    if (this.props.view !== 'email') return null;

    const captchaDisplay = this.props.ignoreCaptcha ? 'none' : 'block';

    return (
      <div>
        <img src={ifIcon} className="otp__logo" alt="infinity" />
        <div className="center-block otp__box">
          <h1 className="otp__header">
            <FormattedMessage {...messages.signInTitle} />
          </h1>
          <form onSubmit={this.onEmailSubmit} testid="emailSubmitForm">
            <div className="form-group">
              <FormattedMessage {...messages.emailPlaceholder}>
                {msg => (
                  <input
                    className="form-control otp__input"
                    name="email"
                    type="email"
                    required
                    placeholder={msg}
                    value={email}
                    onChange={this.handleChange}
                  />
                )}
              </FormattedMessage>
            </div>
            <div
              id="div_id_captcha"
              className="form-group"
              style={{ display: captchaDisplay }}
            >
              <div className="row">
                <div className="col-md-8">
                  <input
                    name="captcha_0"
                    type="hidden"
                    value={captcha_0}
                    required
                  />
                  <FormattedMessage {...messages.captchaPlaceholder}>
                    {msg => (
                      <input
                        className="form-control otp__input"
                        value={captcha_1}
                        name="captcha_1"
                        spellCheck="false"
                        placeholder={msg}
                        required
                        onChange={this.handleChange}
                      />
                    )}
                  </FormattedMessage>
                </div>
                <div className="otp__captcha">
                  <img src={captcha.image_url} alt="captcha" />
                  <a
                    className="primaryAction btn otp__btn"
                    onClick={this.refresh}
                  >
                    &#8634;
                  </a>
                </div>
              </div>
            </div>
            <div>
              <CheckBox
                value={tosAgreement}
                error={tosError}
                action={this.handleToS}
              >
                <Fragment>
                  <span onClick={this.handleToS} testid="tosText">
                    <FormattedMessage {...messages.tosRead} />{' '}
                  </span>
                  <Link to="/terms">
                    <FormattedMessage {...messages.tos} />
                  </Link>
                </Fragment>
              </CheckBox>
            </div>
            <button type="submit" className="primaryAction btn btn-lg otp__btn">
              <FormattedMessage {...messages.continue} />
            </button>
          </form>
        </div>
      </div>
    );
  }
}
