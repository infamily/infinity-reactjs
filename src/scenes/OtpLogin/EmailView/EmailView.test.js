import React from 'react';
import ReactDOM from 'react-dom';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import EmailView from './index';

configure({ adapter: new Adapter() });

describe('<EmailView />', () => {
  it('should render <EmailView /> components', () => {
    const wrapper = shallow(<EmailView />);
  });
});
