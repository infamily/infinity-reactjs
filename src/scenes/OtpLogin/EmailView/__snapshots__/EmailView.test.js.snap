// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EmailView /> matches snapshot 1`] = `
<BrowserRouter
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": Object {},
      "now": [Function],
      "onError": [Function],
      "textComponent": "span",
      "timeZone": null,
    }
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <EmailView
      changeEmail={
        [MockFunction] {
          "calls": Array [
            Array [
              "tony@stark.com",
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        }
      }
      changeView={[MockFunction]}
      setPopUp={[MockFunction]}
      view="email"
    >
      <div>
        <img
          alt="infinity"
          className="otp__logo"
          src="if.png"
        />
        <div
          className="center-block otp__box"
        >
          <h1
            className="otp__header"
          >
            <FormattedMessage
              defaultMessage="Sign In"
              id="infinity.scenes.OtpLogin.signInTitle"
              values={Object {}}
            >
              <span>
                Sign In
              </span>
            </FormattedMessage>
          </h1>
          <form
            onSubmit={[Function]}
            testid="emailSubmitForm"
          >
            <div
              className="form-group"
            >
              <FormattedMessage
                defaultMessage="E-mail address"
                id="infinity.scenes.OtpLogin.EmailView.emailPlaceholder"
                values={Object {}}
              >
                <input
                  className="form-control otp__input"
                  name="email"
                  onChange={[Function]}
                  placeholder="E-mail address"
                  required={true}
                  type="email"
                  value=""
                />
              </FormattedMessage>
            </div>
            <div
              className="form-group"
              id="div_id_captcha"
              style={
                Object {
                  "display": "block",
                }
              }
            >
              <div
                className="row"
              >
                <div
                  className="col-md-8"
                >
                  <input
                    name="captcha_0"
                    required={true}
                    type="hidden"
                    value=""
                  />
                  <FormattedMessage
                    defaultMessage="Captcha"
                    id="infinity.scenes.OtpLogin.EmailView.captchaPlaceholder"
                    values={Object {}}
                  >
                    <input
                      className="form-control otp__input"
                      name="captcha_1"
                      onChange={[Function]}
                      placeholder="Captcha"
                      required={true}
                      spellCheck="false"
                      value=""
                    />
                  </FormattedMessage>
                </div>
                <div
                  className="otp__captcha"
                >
                  <img
                    alt="captcha"
                    src=""
                  />
                  <a
                    className="primaryAction btn otp__btn"
                    onClick={[Function]}
                  >
                    â†º
                  </a>
                </div>
              </div>
            </div>
            <div>
              <CheckBox
                action={[Function]}
                error={false}
                value={false}
              >
                <div
                  className="check_box__container "
                >
                  <div
                    className="check_box "
                    onClick={[Function]}
                  />
                  <p>
                    <small>
                      <span
                        onClick={[Function]}
                        testid="tosText"
                      >
                        <FormattedMessage
                          defaultMessage="I have read and agree to the"
                          id="infinity.scenes.OtpLogin.EmailView.tosRead"
                          values={Object {}}
                        >
                          <span>
                            I have read and agree to the
                          </span>
                        </FormattedMessage>
                         
                      </span>
                      <Link
                        replace={false}
                        to="/terms"
                      >
                        <a
                          href="/terms"
                          onClick={[Function]}
                        >
                          <FormattedMessage
                            defaultMessage="Terms of Service"
                            id="infinity.scenes.OtpLogin.EmailView.tos"
                            values={Object {}}
                          >
                            <span>
                              Terms of Service
                            </span>
                          </FormattedMessage>
                        </a>
                      </Link>
                    </small>
                  </p>
                </div>
              </CheckBox>
            </div>
            <button
              className="primaryAction btn btn-lg otp__btn"
              type="submit"
            >
              <FormattedMessage
                defaultMessage="Continue"
                id="infinity.scenes.OtpLogin.EmailView.continue"
                values={Object {}}
              >
                <span>
                  Continue
                </span>
              </FormattedMessage>
            </button>
          </form>
        </div>
      </div>
    </EmailView>
  </Router>
</BrowserRouter>
`;
